nbuild/stdenv/fetch.py

```diff
@@ -24,42 +22,61 @@ def fetch_url(
 ):
     package = get_package()
 
+    url_object = urlparse(url)
     path = os.path.join(
         package.download_dir,
-        os.path.basename(urlparse(url).path)
+        os.path.basename(url_object.path)
     )
 
     if not sha256:
         wlog(f"No sha256 to ensure the integrity of {url}")
 
     if not _check_file(path, md5, sha1, sha256):
         ilog(f"Fetching {url}")
-        req = requests.get(url, stream=True)
-        with open(path, 'wb') as file:
-            for chunk in req.iter_content(chunk_size=4096):
-                file.write(chunk)
+        if url_object.scheme == 'http' or url_object.scheme == 'https':
+            download_http(url, path)
+        elif url_object.scheme == ('ftp'):
+            download_ftp(url_object, path)
+        else:
+            flog(f"Unknown protocol to download file from url {url}")
+            exit(1)
         clog(f"Fetch done. Stored at {path}")
         if not _check_file(path, md5, sha1, sha256):
             flog(
                 "Downloaded file's signature is invalid. "
                 "Please verify the signature(s) in the build manifest "
                 "and the authenticity of the given link."
             )
-            print(f"Url: {url}\nmd5: {md5}")
             exit(1)
     else:
         clog(f"Using cache at {path}")
 
 
+def download_http(url, path):
+        req = requests.get(url, stream=True)
+        with open(path, 'wb') as file:
+            for chunk in req.iter_content(chunk_size=4096):
+                file.write(chunk)
+
+
+def download_ftp(url_object, path):
+    ftp = ftplib.FTP(url_object.netloc)
+    ftp.login()
+    with open(path, 'wb') as out_file:
+        ftp.retrbinary(f'RETR {url_object.path}',
```

Comment:
We didn't talked about this before, but multi-lines calls should look like this:

```python
my_func(
    param1,
    param2,
    param3,
)
```
(Don't miss the last comma)

It makes it easy to add/remove parameters or swap them.

Reply:
Ok, done.
You might want to create a guide for these kind of rules, if there are any others, and for other contributors

Reply:
Yeah we should, but honestly haven't found the time yet :/